### ISSUE TITLE
`Show` instance for `Compose`

### ISSUE THREAD

**Issue by andrewthad** [VinylRecords/Vinyl#72]
The `Compose` in [transformers](http://hackage.haskell.org/package/transformers-0.4.3.0/docs/Data-Functor-Compose.html#t:Compose) has a `Show` instance`. Can we get that for the one in `vinyl`? Also, I remember reading somewhere an explanation of why vinyl has duplicates on some of the `Data.Functor` types, but I can't recall why that is. Can the reasons be added to the Haddocks?

**Comment by jonsterling** [comment]
The reason for the duplicate functors is that ours are kind-polymorphic, but the standard ones are fixed at `*`. 
- [ ] we can certainly explain this in the haddocks
- [ ] I think we should also be able to get a `Show` instance

I would certainly merge pull requests that did either of these things, or I can also do it (though it may take a bit longer if I have to do it).

Thank you very much for opening this ticket!

**Comment by tel** [comment]
It ought to be reasonable to ask on `libraries` for `transformers` to be kind polymorphic, no?

**Comment by andrewthad** [comment]
@tel I have sent an email to Ross Paterson about this. Unfortunately, transformers in on darcshub, so I don't know what the normal procedure for opening tickets is.
@jonsterling I'll wait for Ross's reply, and if transformers can't be changed, then I'll put together a PR for this. (I am assuming that you would be willing to depend on transformers if the data types were changed. Also, we'd get the `Show` instance for free if it worked out).

**Comment by jonsterling** [comment]
@tel Yes, I think that's reasonable!

@andrewthad Thanks! I think we'd probably be willing to depend on `transformers` since it's pretty standard, though let's just ask @acowley to be sure. It *is* nice that Vinyl is totally immune to Haskell-community library churn at the moment, but it is kind of confusing that we duplicate certain types. (My philosophy is that for types which are internal, they should be inlined; but we expose APIs for these types, so it makes sense that they should be composable with code that uses the standard ones.)

**Comment by acowley** [comment]
I'm more of a "let libraries multiply and flourish" kind of person than Jon, but he's actually convinced me of how valuable some amount of insulation from external dependencies can be. I've actually never run into a problem with incompatibility between `transformers` and `Vinyl`'s internal definitions of those simple types. So, all told, I'm neutral to slightly negative on the `transformers` dependency, but I'd change my mind if somebody came across a use case where direct compatibility was really helpful.

With `transformers` in particular, I'm glad to have not had Vinyl add to the headaches of `ExceptT` and the need for `transformers-compat`. While `transformers` has a venerable history, it's provided a very bumpy ride for the past year.

**Comment by andrewthad** [comment]
Fair enough. Right now, I don't have a compelling reason to reuse the ones from transformers. It seems like few libraries (except for everything Edward Kmett does) use `Compose`, `Const`, etc., so I'm content to wait and see if a compelling use case comes up.

Just to make sure this is noted, I will mention one small benefit of leaning on `transformers` (too small in my opinion to be compelling though). If you look in [Data.Functor.Classes](https://hackage.haskell.org/package/transformers-0.4.3.0/docs/Data-Functor-Classes.html), you'll find a lot of interesting machinery (`Eq1`, `Show1`, `Ord1`, etc.) for helping create instances for the functor newtypes. So the `Eq` instance for `Compose` in [Data.Functor.Compose](https://hackage.haskell.org/package/transformers-0.4.3.0/docs/Data-Functor-Compose.html) looks like this:

    instance (Functor f, Eq1 f, Eq1 g, Eq a) => Eq (Compose f g a)

Based on `vinyl`'s `Storable` instances, a hypothetical equivalent in vinyl that avoids duplicating the `Eq1` machinery would probably look like this:

    instance Eq (f (g x)) => Eq (Compose f g a)

The `transformers` approach is neat because it avoids the need for `FlexibleInstances`. So, you would get slightly more sound instances of a bunch of the common typeclasses for free. That's all. But like I said, not compelling enough on its own, but maybe something to keep in mind for the future.

**Comment by jonsterling** [comment]
I too like the higher-kinded things like `Show1, Eq1`, etc. One thing that I have needed in the past, though, was a heterogeneous `Eq1` type, which let me equate objects in potentially different fibers. That is, to affirm an equality between object of type `f i` with an object of type `f j`, I additionally provide a proof that `i :=: j`. I use this to great effect in `abt`: http://hackage.haskell.org/package/abt-0.1.1.0/docs/Abt-Class-HEq1.html

**Comment by andrewthad** [comment]
The trivial solution (which I just did locally) is to add these lines:

    deriving instance Show a => Show (Const a b)
    deriving instance Show (f (g a)) => Show (Compose f g a)

I added the one for `Const` in their because that one is also helpful for debugging.
